### PR TITLE
Add RDF Canonicalization support

### PR DESCRIPTION
**Pull request by zotanmew**
This PR adds support for [RDF canonicalization](w3c-ccg.github.io/rdf-dataset-canonicalization/spec/), as well as the respective [test suite](https://w3c.github.io/rdf-canon/tests/).

Tasks remaining:
- [x] Fix test 60c (unicode/encoding)
- [x] Cleanup code
- [x] Integrate with JSON-LD handlers
- [x] Fix handling of JSON-LD datetime objects
- [x] Fix handling of URI nodes with a trailing slash in the path that also have an anchor

If I should refactor this to make it more in like with the rest of the codebase, please advise me on which patterns I've missed.

### CODE REVIEW

**Comment by zotanmew** [comment]
Thanks to @deviant’s help, all 84 tests are now passing, and a bunch of bugs in the existing codebase were fixed as well :)

File: Libraries/dotNetRdf.Core/Writing/Formatting/NTriplesFormatter.cs [hunk 1]
```diff
@@ -271,7 +277,19 @@ public override string FormatUri(Uri u)
...
 
-            return FormatUri(Rfc3987Formatter.EscapeUriString(u.ToString()));
+            var uriString = u.ToString();
+            if (uriString.EndsWith("/") && !u.OriginalString.EndsWith("/"))
+                uriString = uriString.Substring(0, uriString.Length - 1);
+            else if (!uriString.EndsWith("/") && u.OriginalString.EndsWith("/"))
+                uriString += '/';
+            else if (uriString.Contains("/#") && !u.OriginalString.Contains("/#"))
+                uriString = uriString.Replace("/#", "#");
+            else if (uriString.Contains("/#") && !u.OriginalString.Contains("/#"))
```
Comment (kal) [inline Libraries/dotNetRdf.Core/Writing/Formatting/NTriplesFormatter.cs:287]: This looks like a duplicate of lines 285-286

**Review by kal — CHANGES_REQUESTED** [review]
Hi @zotanmew and @deviant. Apologies for the delay in getting this PR reviewed! Thanks very much for this contribution. It all looks great and I really appreciate that you added the full test suite too. I just have one minor change request. I'm going to approve the CI run too now

File: Libraries/dotNetRdf.Core/Writing/Formatting/NTriplesFormatter.cs (hunk 1 above)
Comment (zotanmew) [reply Libraries/dotNetRdf.Core/Writing/Formatting/NTriplesFormatter.cs]: Oh, oops, pasted that block twice ^^;

Should be fixed.

**Comment by zotanmew** [comment]
I just realized that the official version of the specification is hosted at https://www.w3.org/TR/rdf-canon/#dfn-rdfc-1-0, while I implemented it based on https://w3c-ccg.github.io/rdf-dataset-canonicalization/spec/. The algorithm they specify is the same, but the former is much more readable. I'm gonna have another look to make sure this is compliant with the actual spec, despite the tests passing.

Also, any opinions on the exposed public methods for this? Are they good as-is? Is it fine that this requires passing an already expanded document? (Seemed like the best thing to do given software might want to process the expanded document on top of getting it canonicalized, to not have to do the work twice)

**Comment by zotanmew** [comment]
I've been referred to https://www.w3.org/TR/rdf-canon/#changes-from-fpwd, which does have changes (albeit minor ones) from the version I implemented. I shall get to work.

**Comment by kal** [comment]
@zotanmew I think it is fine to have the API require an expanded document as input. Convenience methods could always be added later if there is a need. There are a couple of fixes needed to the formatter classes to avoid regressions. I've got the fixes locally but don't have permission to push them to your fork. I'll attach them as a unified diff file here.
[formatter_fixes.patch.gz](https://github.com/dotnetrdf/dotnetrdf/files/13998514/formatter_fixes.patch.gz)

**Comment by zotanmew** [comment]
Thank you so much! I'll finish up the correct algo implementation (now with comments, revised first degree quads pass & compliant return types), merge your patch and then push all of that :)

**Comment by zotanmew** [comment]
All done :)

**Review by kal — APPROVED** [review]
All looks great to me, thank you @zotanmew and @deviant !

**Comment by zotanmew** [comment]
I think the workflow isn't loading the new tests, did I forget something or is this in the workflow definition? (regardless, they all pass on my machine)

**Comment by zotanmew** [comment]
Hmm, one of them crashed o.O (https://github.com/dotnetrdf/dotnetrdf/actions/runs/7595621014/job/20688550712)

**Comment by kal** [comment]
Its in the workflow definition in build.yaml. There's a list of the test projects to run for both the `test` job and the `test-linux` job. Ca you add the test suite project there?

**Comment by kal** [comment]
> Hmm, one of them crashed o.O (https://github.com/dotnetrdf/dotnetrdf/actions/runs/7595621014/job/20688550712)

Yeah there are intermittent errors with the full text query tests. I *think* it may be a race condition with some of the other tests but haven't had time to get to the bottom of it. I've triggered a re-run as that can usually clear it.

**Comment by zotanmew** [comment]
Hm, same test again :/
Since it's crashing it's unlikely to be a regression caused by this, right?

**Comment by kal** [comment]
Yep same build issue as before. I've created a ticket to remind myself to look into this at some point because its really annoying :laughing: 

All checks are passed, so I'll merge now.
Awesome work, thank you very much! :heart:

**Comment by zotanmew** [comment]
Thank you so much! This should probably close #615 as well :)

**Comment by kal** [comment]
> Thank you so much! This should probably close #615 as well :)

Definitely/! I'll write a little bit of documentation for the feature tomorrow and then close the ticket once that is done. :)

**Pull request merged**